Blocks: add paid label next to block title when it requires a plan

jeherve opened the pull request:
#### Changes proposed in this Pull Request:

Related: #11875, #13490 

* This adds a ` (paid)` label next to block titles whenever a block requires a plan.

<img width="439" alt="screenshot 2020-02-19 at 22 38 15" src="https://user-images.githubusercontent.com/426388/74879122-cb419b00-5368-11ea-9314-c4925c1e2843.png">

A few notes:

- I have chosen to have `(beta)` take priority whenever a block is both paid and in Beta. I am not sure that's the best approach.
- (beta) is not translatable. I don't know if it should be.

@jeryj Do you think you could take a look and take over?

Thank you!

#### Testing instructions:

* Start from a site using the free version of Jetpack, and connected to WordPress.com.
* Add the following filter to enable paid blocks on your site: `add_filter( 'jetpack_block_editor_enable_upgrade_nudge', '__return_true' );`
* Open the block picker.

#### Proposed changelog entry for your changes:

* Blocks: better differentiate paid blocks from free ones.

jeherve committed:
Blocks: add paid label next to block title when it requires a plan

jeryj commented:
Thank you for getting this started and providing direction, @jeherve! I have a questions that is likely just related to my newness to Jetpack. The only paid plan block I'm seeing is Recurring Payments, and it's not showing the `(paid)` label. 

This is for both my local docker ngrok'd site and the PR-generated site. They're both connected to my WP account via a free Jetpack plan. I've looked in both wp-admin and WordPress.com's editor for more beta and upgrade-required blocks. Am I looking in the wrong area, or is there some other set-up I need to do to see more beta and paid blocks like in your screenshot?

jeherve commented:
I've added a note to the PR description: to see paid blocks when you don't have a plan in Jetpack, you need to use this filter:
`add_filter( 'jetpack_block_editor_enable_upgrade_nudge', '__return_true' );`

Beta blocks, on the other hand, can be enabled by adding the following to your site's `wp-config.php` file:
`define( 'JETPACK_BETA_BLOCKS', true );`

WunderBart commented:
Thanks for opening this PR and for the detailed description, @jeherve! 🙌 


> I have chosen to have (beta) take priority whenever a block is both paid and in Beta. I am not sure that's the best approach.

How about joining those tags/badges/labels? We'd get something like:

![image](https://user-images.githubusercontent.com/1451471/74934645-fa96ed00-53e6-11ea-8fd0-2399cf669146.png)
 (not sure about the order here ☝️)

> (beta) is not translatable. I don't know if it should be.

I'm not sure here either, but when I drop it in Google Translate, e.g. the Russian version is "бета". I guess the question is if we want to make it translatable?

---

For the implementation of the above, how about creating `blockTags` object that would contain (translated?) strings, e.g.:

```javascript
const blockTags = {
  paid: __( 'paid' ),
  beta: __( 'beta' ),
};
```
...and append (& join) them accordingly to the block's name when needed. What do you think? 
cc @davemart-in

davemart-in commented:
Thanks for kicking this off @jeherve! Combining beta and paid when appropriate sounds fine.

WunderBart committed:
Rename the plan checker func & add comments

WunderBart committed:
Join block tags and make 'beta' translatable

jeherve commented on extensions/shared/register-jetpack-block.js:
```diff
@@ -12,9 +13,21 @@ import getJetpackExtensionAvailability from './get-jetpack-extension-availabilit
 import wrapPaidBlock from './wrap-paid-block';
 
+const availableBlockTags = {
+	paid: __( 'paid', 'jetpack' ),
```
I wonder if we should add some context here with `_x`, to help translators? Something like "Short label appearing near a block requiring a paid plan."

simison replied:
The context would be great. 👍  Otherwise, someone could translate it as "they've already paid for something" vs "this block requires paid plan"

jeherve commented on extensions/shared/register-jetpack-block.js:
```diff
@@ -44,9 +57,23 @@ export default function registerJetpackBlock( name, settings, childBlocks = [] )
+	}
+
+	if ( blockTags.length ) {
+		blockTitle = blockTitle + ` (${ blockTags.join( ', ' ) })`;
```
How about using template literals all the way to stay consistent?

```suggestion
		blockTitle = `${ blockTitle } (${ blockTags.join( ', ' ) })`;
```

simison commented on extensions/shared/register-jetpack-block.js:
```diff
@@ -12,9 +13,21 @@ import getJetpackExtensionAvailability from './get-jetpack-extension-availabilit
 
+const availableBlockTags = {
+	paid: __( 'paid', 'jetpack' ),
+	beta: __( 'beta', 'jetpack' ),
```
Initially, we didn't translate this since it was just internal for a8c only, but not a biggie to translate it either.

jeryj replied:
I've left this translatable for now, but I'm happy to remove it from translation if you'd prefer!

jeryj committed:
Using _x to translate 'paid' string to provide context for the translation

jeryj committed:
Use template literal for building block tags to make the code more readable.

jeryj committed:
Changed the object property from availableBlockTags.plan to availableBlockTags.paid

The availableBlockTags object has the property paid, and not plan. The check to add the (paid) label was looking for the availableBlockTags.plan property, which was undefined. I updated this to look for availableBlockTags.paid.

jeryj commented:
I made the requested changes to the code. I think this is ready for another round of review 👍

getdave commented on extensions/shared/register-jetpack-block.js:
```diff
@@ -44,9 +57,23 @@ export default function registerJetpackBlock( name, settings, childBlocks = [] )
 	}
 
+	let blockTitle = settings.title;
+	const blockTags = [];
```
Minor: it might seem obvious when you've been working on it for a while, but I wonder whether it could be helpful to add a comment here detailing what we're doing here. It was immediately obvious to me even though I had some context. For others it could be even less clear.

For example:

> Some Blocks should be "tagged" to highlight various attributes such as being part of a paid plan or being in a beta programme. We add these to the Block's visible `title` in order that they show up in the picker.

Alternatively, perhaps it would be better to handle adding "tags" in a dedicated method. This might remove the need to document with a comment as the code would be self describing. Perhaps:

`buildBlockTags` and `buildBlockTitle`?

jeryj replied:
I've refactored the implementation here: https://github.com/Automattic/jetpack/pull/14739/commits/25129a6679a36b5f715f9fd7eeb86994d164e1cc

This adds `buildBlockTags` and `buildBlockTitle` functions and builds the title with `buildBlockTitle( settings.title, buildBlockTags( name, requiredPlan ) )`

getdave reviewed (commented):
Nice job. Looks good from a code perspective. I haven't been able to test manually yet. Still working on env setup.

jeherve commented on extensions/shared/register-jetpack-block.js:
```diff
@@ -44,9 +57,23 @@ export default function registerJetpackBlock( name, settings, childBlocks = [] )
+	}
+
+	if ( blockTags.length ) {
+		blockTitle = `${ blockTitle } (${ blockTags.join( ', ' ) })`;
```
@akirk I'd be curious about your opinion about this. Do you think it would be useful to make a string like this translatable? I'm not sure something like "Simple Payments block (paid, beta)" is something that's always going to follow that format in every language. Maybe we should use placeholders in a translatable string here?

jeryj replied:
Same code, but it now lives here: https://github.com/Automattic/jetpack/pull/14739/files#diff-bb769916a770ad52c514ba7ab6609e48R66

Just pinging here since the code here isn't really "Outdated", as github suggests.

akirk replied:
Hard to tell. From a translator perspective a string `%1$s (%2$s)` is always more confusing than helpful, and I'd only go that route when we get complaints that something looks bad. That said, I wonder if we don't have other options from a UI perspective to add "real" badges there vs doing title modification.

jeryj replied:
> I wonder if we don't have other options from a UI perspective to add "real" badges there vs doing title modification.

That was our preferred route, but, sparing you the details, it wasn't going to be allowed to use badges 😞 This `(tags, list)` implementation was what we need to go with for now.

jeryj committed:
Refactors block title and block tags implementation to be into more isolated build functions.

- Adds buildBlockTags( name, requiredPlan)
- Adds buildBlockTitle( blockTitle, blockTags )
- Builds the title in registerJetpackBlock with buildBlockTitle( settings.title, buildBlockTags( name, requiredPlan ) )

getdave commented on extensions/shared/register-jetpack-block.js:
```diff
@@ -12,15 +13,62 @@ import getJetpackExtensionAvailability from './get-jetpack-extension-availabilit
+ */
+function buildBlockTitle( blockTitle, blockTags = [] ) {
+	if ( blockTags.length ) {
+		blockTitle = `${ blockTitle } (${ blockTags.join( ', ' ) })`;
```
```suggestion
		blockTitle = `${ blockTitle } (${ blockTags.join() })`;
```

 If omitted, the array elements are separated with a comma

jeryj replied:
True, but there's no space between them, so it would end up being `(paid,beta)` instead of `(paid, beta)`

WunderBart replied:
They will miss that single whitespace after the comma, though!

getdave reviewed (commented):
Looks much better. Thanks for the refactor 👍 

I'm still setting up my env for testing so please can you source another manual review?

I'm sorry about this.

WunderBart reviewed (approved):
Tested locally, everything looks great! :shipit:

jeryj commented:
I refactored and I believe I addressed all the raised comments. Could I please get another review? 🙏

jeherve commented:
r203949-wpcom
